arc_solvers/nn/util.py

```diff
@@ -32,3 +35,130 @@ def masked_mean(tensor, dim, mask):
+        seq_lens = torch.sum(mask, dim=dim)  # this returns batch_size, 1
+        seq = seq / seq_lens.view([-1, 1])
+
+    return seq
+
+
+def embedd_encode_and_aggregate_text_field(question: Dict[str, torch.LongTensor],
+                                           text_field_embedder,
+                                           embeddings_dropout,
+                                           encoder,
+                                           aggregation_type,
+                                           get_last_states=False):
+    """
+    Given a batched token ids (2D) runs embeddings lookup with dropout, context encoding and aggregation
+    :param question:
+    :param text_field_embedder: The embedder to be used for embedding lookup
+    :param embeddings_dropout: Dropout
+    :param encoder: Context encoder
+    :param aggregation_type: The type of aggregation - max, sum, avg, last
+    :param get_last_states: If it should return the last states.
+    :return:
+    """
+    embedded_question = text_field_embedder(question)
+    question_mask = get_text_field_mask(question).float()
+    embedded_question = embeddings_dropout(embedded_question)
+
+    encoded_question = encoder(embedded_question, question_mask)
+
+    # aggregate sequences to a single item
+    encoded_question_aggregated = seq2vec_seq_aggregate(encoded_question, question_mask, aggregation_type,
+                                                        None, 1)  # bs X d
+
+    last_hidden_states = None
+    if get_last_states:
+        last_hidden_states = allennlp.nn.util.get_final_encoder_states(encoded_question, question_mask, encoder.is_bidirectional())
+
+    return encoded_question_aggregated, last_hidden_states
+
+
+def embedd_encode_and_aggregate_list_text_field(texts_list: Dict[str, torch.LongTensor],
+                                                text_field_embedder,
+                                                embeddings_dropout,
+                                                encoder: Seq2SeqEncoder,
+                                                aggregation_type,
+                                                init_hidden_states=None):
+    """
+    Given a batched list of token ids (3D) runs embeddings lookup with dropout, context encoding and aggregation on
+    :param question:
+    :param text_field_embedder: The embedder to be used for embedding lookup
+    :param embeddings_dropout: Dropout
+    :param encoder: Context encoder
+    :param aggregation_type: The type of aggregation - max, sum, avg, last
+    :param get_last_states: If it should return the last states.
+    :param texts_list:
+    :param text_field_embedder:
+    :param embeddings_dropout:
+    :param encoder:
+    :param aggregation_type:
+    :param init_hidden_states:
+    :return:
+    """
+    embedded_texts = text_field_embedder(texts_list)
+    embedded_texts = embeddings_dropout(embedded_texts)
+
+    bs, ch_cnt, ch_tkn_cnt, d = tuple(embedded_texts.shape)
```

Comment:
Use full variable names here, for readability.

Reply:
Fixed